theano/tensor/opt.py

```diff
@@ -5841,3 +5841,26 @@ def local_add_mul_fusion(node):
 def local_grad_clip(node):
     if isinstance(node.op, theano.gradient.GradClip):
         return node.inputs
+
+@register_canonicalize
+@register_stabilize
+@register_specialize
+@gof.local_optimizer([T.Alloc])
+def local_reduce_double_alloc(node):
```

Comment:
reduce most of the time mean reduction. So I try to find something else. If you have idea, tell them.

What about: local_merge_alloc? Or local_fusion_alloc? I prefer the first, but this can be discussed.